src/Equinox.Core/Caching.fs

```diff
@@ -7,62 +7,52 @@ open System.Diagnostics
 open System.Threading
 open System.Threading.Tasks
 
-type IReloadableCategory<'event, 'state, 'context> =
-    inherit ICategory<'event, 'state, 'context>
-
+type IReloadable<'state> =
     abstract Reload: log: ILogger * streamName: string * requireLeader: bool * streamToken: StreamToken * state: 'state * ct: CancellationToken
                      -> Task<struct (StreamToken * 'state)>
 
-type private Decorator<'event, 'state, 'context>
-    (   category: IReloadableCategory<'event, 'state, 'context>,
-        tryReadCache: string -> Task<voption<struct (StreamToken * 'state)>>,
-        updateCache: string -> struct (StreamToken * 'state) -> Task<unit>) =
-    let cache streamName (inner: CancellationToken -> Task<struct (StreamToken * 'state)>) ct = task {
+type private Decorator<'event, 'state, 'context, 'cat when 'cat :> ICategory<'event, 'state, 'context> and 'cat :> IReloadable<'state> >
```

Comment:
@nordfjord this was what I was trying to do in a rush the other night, based on your distaste of hierarchy. As the PR as a whole has a slight conflict with #387, feel free to review this, and I'll cherry pick this to master once you've looked, with a view to releasing that when you're ready